File: leetcode/python/003_longest_substring_without_repeating_characters.py
# Given a string, find the length of the longest substring without repeating characters.

# Example 1:

# Input: "abcabcbb"
# Output: 3 
# Explanation: The answer is "abc", with the length of 3. 
# Example 2:

# Input: "bbbbb"
# Output: 1
# Explanation: The answer is "b", with the length of 1.
# Example 3:

# Input: "pwwkew"
# Output: 3
# Explanation: The answer is "wke", with the length of 3. 
#              Note that the answer must be a substring, "pwke" is a subsequence and not a substring.


class Solution:

    def lengthOfLongestSubstring(self, s: str) -> int:
        pass

    # 方法一暴力求解
    def lengthOfLongestSubstring_1(self, s: str) -> int:
        if not s: return 0
        m = 0
        for i in range(0, len(s)):
            for j in range(i+1, len(s)+1):
                if len(set(s[i:j])) == j-i:
                    m = max(m, j-i)
        return m

    # 滑动窗口
    def lengthOfLongestSubstring_2(self, s: str) -> int:
        pass

    # 优化滑动窗口
    def lengthOfLongestSubstring_3(self, s: str) -> int:
        pass


if __name__ == "__main__":
    a = "abcabcbb"
    b = "bbbbb"
    c = "pwwkew"
    d = " "
    e = "e"
    solution = Solution()

    # 暴力求解
    assert solution.lengthOfLongestSubstring_1(a) == 3
    assert solution.lengthOfLongestSubstring_1(b) == 1
    assert solution.lengthOfLongestSubstring_1(c) == 3
    assert solution.lengthOfLongestSubstring_1(d) == 1
    assert solution.lengthOfLongestSubstring_1(e) == 1

File: leetcode/go/003/003longest_substring_without_repeating_characters.go
package main

type void struct{}

var null void

func lengthOfLongestSubstring(s string) int {
	return len(s)
}

// 方法一： 暴力求解，时间复杂度为O(n^3)
func lengthoflongestsubstring1(s string) int {
	length := len(s)
	if length == 0 {
		return 0
	}
	maxLength := 0
	for i := 0; i < length; i++ {
		for j := i + 1; j <= length; j++ {
			if isUniqueString(s[i:j]) && maxLength < j-i {
				maxLength = j - i
			}
		}
	}
	return maxLength
}

func isUniqueString(s string) bool {
	set := make(map[string]void)
	for _, i := range s {
		key := string(i)
		if _, exists := set[key]; exists {
			return false
		}
		set[key] = null
	}
	return true
}
